### PR TITLE
Pig: Fix for complex schema conversion

### PR DESCRIPTION
**Pull request by danielcweeks**
Fixes nested schema conversion for complex types.

### CODE REVIEW

File: pig/src/main/java/com/netflix/iceberg/pig/SchemaUtil.java [hunk 1]
```diff
@@ -64,9 +60,25 @@ private static ResourceFieldSchema convert(Type type) throws IOException {
     ResourceFieldSchema result = new ResourceFieldSchema();
     result.setType(convertType(type));
 
+    if(!type.isPrimitiveType()) {
+      ResourceSchema nestedSchema = new ResourceSchema();
+      nestedSchema.setFields(convertFields(type.asNestedType().fields()));
```
Comment (rdsr) [inline pig/src/main/java/com/netflix/iceberg/pig/SchemaUtil.java]: If the input type is a list[long] say, then we will create a bag and add a element [of type long] to it without first wrapping it inside a tuple.

I confirmed this behavior using this test case
` @Test
  public void testLongInBag() throws IOException {
    Schema icebergSchema = new Schema(
        optional(
            1, "nested_list",
            MapType.ofOptional(
                2, 3,
                StringType.get(),
                ListType.ofRequired(5, LongType.get()))));
    SchemaUtil.convert(icebergSchema);
  }`

Also, it seems that this functionality of converting complex iceberg schema to Pig schema  is being done in 2 places. Once in `convert(com.netflix.iceberg.types.Type)` and the other in `convertComplex(Types.NestedField field)` .  Is there scope in simplifying this?

File: pig/src/test/java/com/netflix/iceberg/pig/SchemaUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            MapType.ofOptional(
+                2, 3,
+                StringType.get(),
+                ListType.ofOptional(
+                    4, StructType.of(
+                        required(5, "id", LongType.get()),
+                        optional(6, "data", StringType.get()))))));
+
+    ResourceSchema pigSchema = SchemaUtil.convert(icebergSchema);
+    assertEquals("nested_list:[{element:(id:long,data:chararray)}]", pigSchema.toString()); // The output should contain a nested struct within a list within a map, I think.
```
Comment (rdblue) [inline pig/src/test/java/com/netflix/iceberg/pig/SchemaUtilTest.java]: @rdsr, can you confirm that this is correct?

File: pig/src/main/java/com/netflix/iceberg/pig/SchemaUtil.java [hunk 3]
```diff
@@ -64,9 +60,25 @@ private static ResourceFieldSchema convert(Type type) throws IOException {
     ResourceFieldSchema result = new ResourceFieldSchema();
     result.setType(convertType(type));
 
+    if(!type.isPrimitiveType()) {
```
Comment (rdblue) [inline pig/src/main/java/com/netflix/iceberg/pig/SchemaUtil.java]: Nit: missing space between `if` and the boolean expression.

File: pig/src/test/java/com/netflix/iceberg/pig/SchemaUtilTest.java (hunk 2 above)
Comment (rdsr) [reply pig/src/test/java/com/netflix/iceberg/pig/SchemaUtilTest.java]: Yes, seems correct to me. I also added a few other tests as part of https://github.com/Netflix/iceberg/pull/58

**Comment by danielcweeks** [comment]
@rdsr Can you take a look at this with the latest tests.  I know there's still the open question around the top-level representation of the tuple in the expected results (e.g. `a:int` vs `(a:int)`), but I think this addresses most of the nesting issues you identified.

File: pig/src/main/java/com/netflix/iceberg/pig/SchemaUtil.java [hunk 4]
```diff
@@ -109,15 +118,21 @@ private static ResourceSchema convertComplex(Types.NestedField field) throws IOE
...
 
-        ResourceSchema innerSchema = new ResourceSchema();
-        innerSchema.setFields(new ResourceFieldSchema[]{convert(listType.elementType())});
+        ResourceFieldSchema [] elementFieldSchemas = new ResourceFieldSchema[]{convert(listType.elementType())};
 
-        //Bag elements are tuples
-        ResourceFieldSchema tupleSchema = new ResourceFieldSchema();
-        tupleSchema.setType(DataType.TUPLE);
-        tupleSchema.setSchema(innerSchema);
+        //Wrap non-struct types in tuples
```
Comment (rdsr) [inline pig/src/main/java/com/netflix/iceberg/pig/SchemaUtil.java]: minor: the comment seems to be for the else block

**Comment by rdsr** [comment]
LGTM overall

File: pig/src/main/java/com/netflix/iceberg/pig/SchemaUtil.java (hunk 4 above)
Comment (danielcweeks) [reply pig/src/main/java/com/netflix/iceberg/pig/SchemaUtil.java]: Moved it.

**Comment by danielcweeks** [comment]
@rdsr Thanks for all the help reviewing this.  We don't have a lot of examples/experience with nested data in pig, so this was a huge help.  Let me know if you want me to update the attribution for the test cases to you before I merge it (I don't know if you care or not, but I'm happy to get your name on there).

**Comment by rdblue** [comment]
Thanks for you help, @rdsr! I'm going to merge this.

**Pull request merged**